libraries/AP_HAL/utility/RingBuffer.cpp

```diff
@@ -192,18 +192,13 @@ uint32_t ByteBuffer::read(uint8_t *data, uint32_t len)
 }
 
 /*
-  return a pointer to a contiguous read buffer
+ * Returns the pointer and size to a contiguous read in the buffer
  */
 const uint8_t *ByteBuffer::readptr(uint32_t &available_bytes)
 {
-    available_bytes = available();
-    if (available_bytes == 0) {
-        return nullptr;
-    }
-    if (head+available_bytes > size) {
-        available_bytes = size - head;
-    }
-    return &buf[head];
+    available_bytes = (head > tail) ? size - head : tail - head;
```

Comment:
every time you use both head and tail you need to think about possible races. Here the window is really small but it exists.  The caller is the read thread which means `tail` could have changed.

Suppose `head > tail` is false and before `tail` is fetched again it is updated by the other thread and wraps. This means that suddenly you would have a negative number (actually a very big unsigned number since the variable is unsigned). The fix would be to copy the variable to stack, similar to what `available()` method does:

```c++
const uint8_t *ByteBuffer::readptr(uint32_t &available_bytes)
{
    uint32_t _tail = tail;
    available_bytes = (head > _tail) ? size - head : _tail - head;
    return available_bytes ? &buf[head] : nullptr;
}
```